### PR TITLE
Improve publisher tool

### PR DESCRIPTION
**Pull request by jdisanti**
Issues discovered while publishing:
- `cargo publish` tries to use all of your CPU cores, so running multiple `cargo publish` in parallel wasn't accomplishing much.
- Stats don't count Smithy runtime packages anymore due to the prefix change.
- Cargo returns an error status code when a crate has already been published, and there isn't a CLI flag to make this idempotent.

This PR resolves these issues and also adds a `--continue-from` option to publish to make it possible to continue publishing from where a failure happened in a previous publish. For example, if throttling occurs while trying to publish aws-smithy-query, then `publisher publish --continue-from aws-smithy-query` can be run to resume from where it left off.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/publisher/src/package.rs [hunk 1]
```diff
@@ -254,6 +289,11 @@ fn batch_packages(packages: Vec<Package>) -> Result<Vec<PackageBatch>> {
...
     }
 
+    // Sort packages within batches so that `--continue-from` work consistently
+    for batch in batches.iter_mut() {
+        batch.sort_by(|a, b| a.handle.cmp(&b.handle));
+    }
+
     // Push the final batch
     if !packages.is_empty() {
         batches.push(packages);
```
Comment (rcoh) [inline tools/publisher/src/package.rs:299]: I think this is a bug? we aren't sorting the final batch

File: tools/publisher/src/package.rs [hunk 2]
```diff
@@ -33,7 +33,7 @@ impl PackageHandle {
 }
 
 /// Represents a crate (called Package since crate is a reserved word).
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
```
Comment (rcoh) [inline tools/publisher/src/package.rs:36]: could just impl Ord for Package

File: tools/publisher/src/package.rs [hunk 3]
```diff
@@ -81,17 +81,19 @@ impl PackageStats {
...
-            } else {
-                warn!("Unrecognized crate name: {}", package.handle.name);
+        for batch in batches {
+            for package in batch {
+                if package.handle.name.starts_with("aws-smithy-") {
+                    stats.smithy_runtime_crates += 1;
+                } else if package.handle.name.starts_with("aws-sdk-") {
+                    stats.aws_sdk_crates += 1;
+                } else if package.handle.name.starts_with("aws-") {
+                    stats.aws_runtime_crates += 1;
```
[the remark this answers is not in the document]
Comment (rcoh) [reply tools/publisher/src/package.rs:93]: you can do `s if s.starts_with(...)` 🤷🏻

File: tools/publisher/src/package.rs [hunk 4]
```diff
@@ -106,9 +108,42 @@ pub async fn discover_package_batches(
...
+    batches: &mut Vec<PackageBatch>,
+    stats: &mut PackageStats,
+) -> Result<(), anyhow::Error> {
+    while !batches.is_empty() {
+        let found = {
+            let first_batch = batches.iter().next().unwrap();
+            first_batch.iter().any(|p| p.handle.name == package_name)
+        };
+        if !found {
+            batches.remove(0);
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply tools/publisher/src/package.rs:132]: Yeah, for this use-case, the 260+ packages get batched instantly, so I don't think I'll bother with optimizing it 😂 

We can revisit if it becomes a problem with more packages.

File: tools/publisher/src/package.rs (hunk 1 above)
Comment (jdisanti) [reply tools/publisher/src/package.rs:299]: Great catch! Fixed and added a test case that repros it.

**Comment by jdisanti** [comment]
Added a few more things since the original PR description:
- Since publishing is so slow, check to see a package is published before trying to publish. This kind of eliminates the need for `--continue-from`, so we could remove that. I've kept it for now.
- Added the ability to yank large groups of package versions so that I could eliminate all the squatted crate versions that were at `0.0.0`.
- Disabled incremental compilation since I definitely ran out of disk space during publishing 😅 
- Added code to add the SDK owner team to the crate ownership. We can have it remove individual user names too. I just haven't put in the effort yet.

File: tools/publisher/fake_cargo/cargo_owner_list [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+>&2 echo "     Updating crates.io index"
+echo "rcoh (Russel Cohen)"
```
Comment (rcoh) [inline tools/publisher/fake_cargo/cargo_owner_list:3]:
```suggestion
echo "rcoh (Russell Cohen)"
```

File: tools/publisher/fake_cargo/cargo_yank_not_found [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+>&2 echo "     Updating crates.io index"
+>&2 echo "        Yank aws-sigv4:0.0.0"
+>&2 echo "error: failed to yank from the registry at https://crates.io"
+>&2 echo
+>&2 echo "Caused by:"
+>&2 echo '  the remote server responded with an error: crate `aws-sigv4` does not have a version `0.0.0`'
```
Comment (rcoh) [inline tools/publisher/fake_cargo/cargo_yank_not_found:7]: if you want, you can use bash anonymous functions to redirect an entire block:
```suggestion
{
  echo "     Updating crates.io index"
  echo "        Yank aws-sigv4:0.0.0"
  echo "error: failed to yank from the registry at https://crates.io"
  echo
  echo "Caused by:"
  echo '  the remote server responded with an error: crate `aws-sigv4` does not have a version `0.0.0`'
} >&2
```

File: tools/publisher/src/cargo.rs [hunk 7]
```diff
@@ -5,82 +5,72 @@
...
+    let message = format!(
+        "Failed to {}:\nStatus: {}\nStdout: {}\nStderr: {}\n",
+        operation_name,
+        if let Some(code) = output.status.code() {
+            format!("{}", code)
+        } else {
+            "Killed by signal".to_string()
+        },
+        String::from_utf8_lossy(&output.stdout),
+        String::from_utf8_lossy(&output.stderr)
```
Comment (rcoh) [inline tools/publisher/src/cargo.rs:73]: when there are a lot of params, named params are a little easier to read IMO:
```suggestion
        "Failed to {name}:\nStatus: {status}\nStdout: {stdout}\nStderr: {stderr}\n",
        name = operation_name,
        status = if let Some(code) = output.status.code() {
            format!("{}", code)
        } else {
            "Killed by signal".to_string()
        },
        stdout = String::from_utf8_lossy(&output.stdout),
        stderr = String::from_utf8_lossy(&output.stderr)
```

File: tools/publisher/src/cargo/check_published.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        let mut command = Command::new(self.program);
+        command
+            .current_dir(self.package_path)
+            .arg("search")
+            .arg(self.package_handle.name.to_string());
+        let output = tokio::task::spawn_blocking(move || command.output()).await??;
+        handle_failure("check published state", &output)?;
+        let (stdout, _) = output_text(&output);
+
+        let line_re = Regex::new(r#"^([a-z0-9\-_]+)\s*=\s*"([a-z0-9\.\-]+)".*$"#).unwrap();
```
Comment (rcoh) [inline tools/publisher/src/cargo/check_published.rs:46]: can you add a comment about what this matches?

File: tools/publisher/src/cargo/get_owners.rs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        .await
+        .unwrap();
+        assert_eq!(
+            vec![
+                "rcoh".to_string(),
+                "github:awslabs:rust-sdk-owners".to_string()
+            ],
+            owners
+        );
+    }
```
Comment (rcoh) [inline tools/publisher/src/cargo/get_owners.rs:79]: this is a really cool idea.

**Review by rcoh — APPROVED** [review]

File: tools/publisher/fake_cargo/cargo_yank_not_found (hunk 6 above)
Comment (jdisanti) [reply tools/publisher/fake_cargo/cargo_yank_not_found:7]: Nice!

**Pull request merged**